hbase-client/src/main/java/org/apache/hadoop/hbase/security/provider/DefaultProviderSelector.java

```diff
@@ -0,0 +1,121 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.security.provider;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.security.User;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@InterfaceAudience.Private
+public class DefaultProviderSelector implements AuthenticationProviderSelector {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultProviderSelector.class);
+  private final ReadWriteLock LOCK = new ReentrantReadWriteLock();
+
+  Configuration conf;
+  SimpleSaslClientAuthenticationProvider simpleAuth = null;
+  GssSaslClientAuthenticationProvider krbAuth = null;
+  DigestSaslClientAuthenticationProvider digestAuth = null;
+
+  @Override
+  public void configure(Configuration conf, Map<Byte,SaslClientAuthenticationProvider> providers) {
+    try {
+      LOCK.writeLock().lock();
+      if (this.conf != null) {
+        throw new IllegalStateException("configure() should only be called once");
+      }
+      this.conf = Objects.requireNonNull(conf);
+    } finally {
+      LOCK.writeLock().unlock();
+    }
+    
+    for (SaslClientAuthenticationProvider provider : Objects.requireNonNull(providers).values()) {
+      if (provider instanceof SimpleSaslClientAuthenticationProvider) {
+        if (simpleAuth != null) {
+          throw new IllegalStateException(
+              "Encountered multiple SimpleSaslClientAuthenticationProvider instances");
+        }
+        simpleAuth = (SimpleSaslClientAuthenticationProvider) provider;
+      } else if (provider instanceof GssSaslClientAuthenticationProvider) {
+        if (krbAuth != null) {
+          throw new IllegalStateException(
+              "Encountered multiple GssSaslClientAuthenticationProvider instances");
+        }
+        krbAuth = (GssSaslClientAuthenticationProvider) provider;
+      } else if (provider instanceof DigestSaslClientAuthenticationProvider) {
+        if (digestAuth != null) {
+          throw new IllegalStateException(
+              "Encountered multiple DigestSaslClientAuthenticationProvider instances");
+        }
+        digestAuth = (DigestSaslClientAuthenticationProvider) provider;
+      } else {
+        LOG.warn("Ignoring unknown SaslClientAuthenticationProvider: {}", provider.getClass());
```

Comment:
Looks like client side doesn't support self-implemented `SaslClientAuthenticationProvider`. `Pluggable`?

Reply:
HBase can't know how/when to use a custom-built authentication token. You provide your own implementation of a `TokenSelector`.